### ISSUE TITLE
sqlite3.Cursor.close() crashes in case the Cursor object is uninitialized

### ISSUE THREAD

**Issue by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [python/cpython#75945]
BPO | [31764](https://bugs.python.org/issue31764)
--- | :---
Nosy | @vstinner, @orenmn
PRs | <li>python/cpython#3958</li><li>python/cpython#4303</li><li>python/cpython#4333</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-11-08.09:59:12.399>
created_at = <Date 2017-10-11.19:09:18.400>
labels = ['extension-modules', '3.7', 'type-crash']
title = 'sqlite3.Cursor.close() crashes in case the Cursor object is uninitialized'
updated_at = <Date 2017-11-08.09:59:12.398>
user = 'https://github.com/orenmn'
```

bugs.python.org fields:
```python
activity = <Date 2017-11-08.09:59:12.398>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-11-08.09:59:12.399>
closer = 'vstinner'
components = ['Extension Modules']
creation = <Date 2017-10-11.19:09:18.400>
creator = 'Oren Milman'
dependencies = []
files = []
hgrepos = []
issue_num = 31764
keywords = ['patch']
message_count = 7.0
messages = ['304172', '305693', '305694', '305701', '305813', '305815', '305816']
nosy_count = 2.0
nosy_names = ['vstinner', 'Oren Milman']
pr_nums = ['3958', '4303', '4333']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'crash'
url = 'https://bugs.python.org/issue31764'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [comment]
The following code causes a crash:
import sqlite3
cursor = sqlite3.Cursor.\_\_new__(sqlite3.Cursor)
cursor.close()

this is because pysqlite_cursor_close() (in [Modules/_sqlite/cursor.c](https://github.com/python/cpython/blob/main/Modules/_sqlite/cursor.c)) assumes
that `self->connection` is not NULL, and passes it to pysqlite_check_thread(),
which crashes.


I would open a PR to fix this soon.

**Comment by vstinner** [comment]
New changeset edb13ae48c17210fa4b2d40a6833ca09db5c121b by Victor Stinner (Oren Milman) in branch 'master':
bpo-31764: Prevent a crash in sqlite3.Cursor.close() in case the Cursor object is uninitialized (bpo-3958)
https://github.com/python/cpython/commit/edb13ae48c17210fa4b2d40a6833ca09db5c121b

**Comment by vstinner** [comment]
@Oren: The 2.7 backport is not straighforward. Would you mind to backport the fix to 2.7, please?

**Comment by vstinner** [comment]
New changeset b0331c94c2a210d50e43d99b249ec83ee165e70c by Victor Stinner (Miss Islington (bot)) in branch '3.6':
bpo-31764: Prevent a crash in sqlite3.Cursor.close() in case the Cursor object is uninitialized (GH-3958) (bpo-4303)
https://github.com/python/cpython/commit/b0331c94c2a210d50e43d99b249ec83ee165e70c

**Comment by 4e331a12-c2bb-4cd1-b83c-dcc766e8fbf8** [comment]
I opened bpo-4333 for 2.7, but it is quite straightforward.. Am i missing something?

**Comment by vstinner** [comment]
New changeset cd66d6d632b4a8190e97f265ee163f11b5ae38e6 by Victor Stinner (Oren Milman) in branch '2.7':
bpo-31764: Prevent a crash in sqlite3.Cursor.close() in case the Cursor object is uninitialized (GH-4333)
https://github.com/python/cpython/commit/cd66d6d632b4a8190e97f265ee163f11b5ae38e6

**Comment by vstinner** [comment]
Thank you for the backport Oren, I already merged it into Python 2.7.

Thanks Oren Milman for your bug report \*and* for your bug fix!


> I opened bpo-4333 for 2.7, but it is quite straightforward.. Am i missing something?

Sorry, I had to say that "git cherry-pick" failed on 2.7. I didn't check why ;-)
